genie-web/src/main/java/com/netflix/genie/web/services/impl/DiskLogServiceImpl.java

```diff
@@ -0,0 +1,336 @@
+            reference and don't rely on properties. Too much to fix right now with other priorities.
+            This should probably be based off the property.
+         */
+
+        // TODO: This throws an InvalidPathException ... Should we catch or just propagate as this will be called
+        //       at startup and the system won't even start
+        this.jobsDirRoot = jobsDir.getFile().toPath();
+
+        // Make sure the root exists and is a directory
+        this.createOrCheckDirectory(this.jobsDirRoot);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void createJobDirectory(final String jobId) throws IOException {
+        // TODO: Throws InvalidPathException...catch and rethrow or let it go? Even Java SDK throws Marco nemesis
+        final Path jobDirectory = this.jobsDirRoot.resolve(jobId);
+        this.createOrCheckDirectory(jobDirectory);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<JobLogFile> getJobDirectoryFileState(final String jobId, final boolean calculateMd5) throws IOException {
+        log.debug("Getting job directory state for job {} {} MD5", jobId, calculateMd5 ? "with" : "without");
+        // TODO: It's possible the system should lock this directory while reading for consistent state?
+        final Path jobDirectory = this.jobsDirRoot.resolve(jobId);
+        this.createOrCheckDirectory(jobDirectory);
+
+        final Set<JobLogFile> jobDirectoryFiles = Sets.newHashSet();
+        Files.walkFileTree(jobDirectory, new FileStateVisitor(jobDirectory, calculateMd5, jobDirectoryFiles));
+
+        return jobDirectoryFiles;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    // TODO: We should be careful about how large the byte[] is. Perhaps we should have precondition to protect memory
+    //       or we should wrap calls to this in something that chunks it off an input stream or just take this in as
+    //       input stream
+    public void updateLogFile(
+        final String jobId,
+        final String relativePath,
+        final long startByte,
+        final byte[] data
+    ) throws IOException {
+        log.debug(
+            "Attempting to write {} bytes from position {} into log file {} for job {}",
+            data.length,
+            startByte,
+            relativePath,
+            jobId
+        );
+        final Path logFile = this.jobsDirRoot.resolve(jobId).resolve(relativePath);
+
+        if (Files.notExists(logFile)) {
+            // Make sure all the directories exist on disk
+            final Path logFileParent = logFile.getParent();
+            if (logFileParent != null) {
+                this.createOrCheckDirectory(logFileParent);
+            }
+        } else if (Files.isDirectory(logFile)) {
+            // TODO: Perhaps this should be different exception
+            throw new IllegalArgumentException(relativePath + " is a directory not a file. Unable to update");
```

Comment:
Should be same exception as the one we decide to throw at line 256

Reply:
This is a user input error.
Line 256 is the unlikely event that someone is creating files inside a private folder that genie is managing.
I disagree they need to be treated the same (even if the underlying error -- file where i'm attempting to create a directory -- is the same).

Reply:
Whoops, missed placed where the other was invoked, it can be more than user error.

Still think on a very abstract level that it can be appropriate for the same underlying error/state to throw different kinds of error depending on context (and example here is user input/configuration vs. runtime unexpected state).

Reply:
The argument could be made that since this service exist only to write files from the `Agent` and provide handles to files to the API tier on `GET` requests that the Genie application completely owns (or should own) this directory and its entire tree. As such the structure should never match these conditions and they're exceptional (e.g. someone logs into the server and creates a file on disk)

One weird edge case could be that currently this service only handles writing but if a job at runtime does file renames/deletions/etc things could get dicey. We probably need to think more about this case. Not sure if the watcher implementation on the Agent is set up to catch these events from the FS or not.